src/plugins/lvm-dbus.c

```diff
@@ -1160,6 +1160,116 @@ static gchar* _lvm_metadata_lv_name (const gchar *vg_name, const gchar *lv_name,
     return g_strstrip (g_strdelimit (ret, "[]", ' '));
 }
 
+static BDLVMSEGdata** _lvm_segs (const gchar *vg_name, const gchar *lv_name, GError **error) {
+    GVariant *prop = NULL;
+    BDLVMSEGdata **segs;
+    gsize n_segs;
+    GVariantIter iter, iter2;
+    GVariant *pv_segs, *pv_name_prop;
+    const gchar *pv;
+    gchar *pv_name;
+    gsize pv_first_pe, pv_last_pe;
+    int i;
+
+    prop = get_lv_property (vg_name, lv_name, "Devices", error);
+    if (!prop)
+        return NULL;
+
+    // Count number of segments
+    n_segs = 0;
+    g_variant_iter_init (&iter, prop);
+    while (g_variant_iter_loop (&iter, "(&o@a(tts))", NULL, &pv_segs))
+      n_segs += g_variant_n_children (pv_segs);
+
+    if (n_segs == 0) {
+      g_variant_unref (prop);
+      return NULL;
+    }
+
+    // Build segments
+    segs = g_new0 (BDLVMSEGdata *, n_segs+1);
+    i = 0;
+    g_variant_iter_init (&iter, prop);
+    while (g_variant_iter_next (&iter, "(&o@a(tts))", &pv, &pv_segs)) {
+      pv_name_prop = get_object_property (pv, PV_INTF, "Name", NULL);
+      if (pv_name_prop) {
+        g_variant_get (pv_name_prop, "&s", &pv_name);
+        g_variant_iter_init (&iter2, pv_segs);
+        while (g_variant_iter_next (&iter2, "(tt&s)", &pv_first_pe, &pv_last_pe, NULL)) {
+          BDLVMSEGdata *seg = g_new0(BDLVMSEGdata, 1);
+          seg->pv_start_pe = pv_first_pe;
+          seg->size_pe = pv_last_pe - pv_first_pe + 1;
+          seg->pvdev = g_strdup (pv_name);
```

Comment:
> The failures on 32bit Debian don't look random. `lvm_dbus_tests.LvmTestLVsMultiSegment` and `lvm_dbus_tests.LvmTestPartialLVs` fail both on stable and testing. It's probably a problem with lvmdbusd, but we should be sure before we skip the tests on 32bit.

So lvmdbusd works correctly, seems like it's glib that doesn't like our code: `pv_name` is suddently `NULL` in the nested while here. I'm not sure why, we get only reference from `pv_name_prop` but it shouldn't suddenly be freed and it didn't went out of scope, right?

Reply:
The bug was to use `gsize` for `pv_first_pe` and `pv_last_pe`. It's only 4 bytes on 32 bit arches, but `g_variant_iter_next` will write 8 bytes for a "t" field. That was overwriting `pv_name` on the stack. I have pushed a fix.

Reply:
Awesome! Haven't really noticed this.